sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ListStateImpl.scala

```diff
@@ -0,0 +1,121 @@
+   override def exists(): Boolean = {
+     val encodedGroupingKey = StateTypesEncoderUtils.encodeGroupingKey(stateName, keyExprEnc)
+     val stateValue = store.get(encodedGroupingKey, stateName)
+     stateValue != null
+   }
+
+   /** Get the state value if it exists. If the state does not exist in state store, an
+    * empty iterator is returned. */
+   override def get(): Iterator[S] = {
+     val encodedKey = StateTypesEncoderUtils.encodeGroupingKey(stateName, keyExprEnc)
+     val unsafeRowValuesIterator = store.valuesIterator(encodedKey, stateName)
+     new Iterator[S] {
+       override def hasNext: Boolean = {
+         unsafeRowValuesIterator.hasNext
+       }
+
+       override def next(): S = {
+         val valueUnsafeRow = unsafeRowValuesIterator.next()
+         StateTypesEncoderUtils.decodeValue(valueUnsafeRow)
+       }
+     }
+   }
+
+   /** Get the list value as an option if it exists and None otherwise. */
+   override def getOption(): Option[Iterator[S]] = {
+     Option(get())
+   }
+
+   /** Update the value of the list. */
+   override def put(newState: Array[S]): Unit = {
+     validateNewState(newState)
+
+     if (newState.isEmpty) {
+       this.remove()
+     } else {
+       val encodedKey = StateTypesEncoderUtils.encodeGroupingKey(stateName, keyExprEnc)
+
+       var isFirst = true
+       newState.foreach { v =>
+         val encodedValue = StateTypesEncoderUtils.encodeValue(v)
+         if (isFirst) {
+           store.put(encodedKey, encodedValue, stateName)
+           isFirst = false
+         } else {
+            store.merge(encodedKey, encodedValue, stateName)
+         }
+       }
+     }
+   }
+
+   /** Append an entry to the list. */
+   override def appendValue(newState: S): Unit = {
+     if (newState == null) {
+       throw new IllegalArgumentException("value added to ListState should be non-null")
+     }
+     store.merge(StateTypesEncoderUtils.encodeGroupingKey(stateName, keyExprEnc),
+         StateTypesEncoderUtils.encodeValue(newState), stateName)
+   }
+
+   /** Append an entire list to the existing value. */
+   override def appendList(newState: Array[S]): Unit = {
+     validateNewState(newState)
+
+     val encodedKey = StateTypesEncoderUtils.encodeGroupingKey(stateName, keyExprEnc)
+     newState.foreach { v =>
+       val encodedValue = StateTypesEncoderUtils.encodeValue(v)
+       store.merge(encodedKey, encodedValue, stateName)
+     }
+   }
+
+   /** Remove this state. */
+   override def remove(): Unit = {
+     store.remove(StateTypesEncoderUtils.encodeGroupingKey(stateName, keyExprEnc), stateName)
+   }
+
+   private def validateNewState(newState: Array[S]): Unit = {
+     if (newState == null) {
+       throw new IllegalArgumentException("newState list should be non-null")
+     }
+
+     val containsNullElements = newState.contains(null.asInstanceOf[S])
+     if (containsNullElements) {
+       throw new IllegalArgumentException("value added to ListState should be non-null")
```

Comment:
Lets move to the NERF framework for this now ?

Reply:
Done.